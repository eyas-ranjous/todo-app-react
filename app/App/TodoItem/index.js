import React from 'react';
import TodoStatusIcon from './TodoStatusIcon';
import RemoveIcon from './RemoveIcon';
import EditIcon from './EditIcon';
import TodoText from './TodoText';

const TodoItem = ({ todo }) => (
  <li>
    <TodoStatusIcon todo={todo} />
    <RemoveIcon todo={todo} />
    <EditIcon todo={todo} />
    <TodoText todo={todo} />
  </li>
);

export default TodoItem;
