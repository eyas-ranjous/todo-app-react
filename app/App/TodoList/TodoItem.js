import React, { useRef } from 'react';
import classnames from 'classnames';

import TodoStatus from '../TodoStatus';

import styles from './styles.css';

export default ({
  todo,
  toggleTodoStatus,
  editTodo,
  unEditTodo,
  saveEditedTodo,
  removeTodo,
  isEditing
}) => {
  const todoInput = useRef();

  const handleEditKeyDown = (e) => {
    if (e.key === 'Enter') {
      const text = todoInput.current.value.trim();
      if (!text) {
        unEditTodo(todo.id);
        return;
      }
      saveEditedTodo({ ...todo, ...{ text } });
    }
  };

  return (
    <li>
      <span
        title="change"
        className={styles.icon}
        onClick={() => toggleTodoStatus(todo)}
      >
        {todo.status === TodoStatus.done ? '✅' : '⬜'}
      </span>

      <span title="remove"
        className={classnames(styles.icon, styles['remove-icon'])}
        onClick={() => removeTodo(todo)}
      >
        ✘
      </span>

      <span
        title="edit"
        className={classnames(styles.icon, styles['edit-icon'])}
        onClick={() => editTodo(todo.id)}
      >
        ✎
      </span>

      {!isEditing && (
        <span className={todo.status === TodoStatus.done
          ? classnames(styles.text, styles.done)
          : styles.text
        }>
          {todo.text}
        </span>
      )}

      {isEditing && (
        <input
          type="text"
          onKeyDown={handleEditKeyDown}
          defaultValue={todo.text}
          ref={todoInput}
        />
      )}
    </li>
  );
};
