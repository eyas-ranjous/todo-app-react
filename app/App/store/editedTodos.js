const editedTodos = (state = new Set(), action) => {
  if (action.type === 'EDIT_TODO') {
    return new Set(state).add(action.todoId);
  }

  if (action.type === 'REMOVE_EDITED_TODO') {
    const newState = new Set(state);
    newState.delete(action.todoId);
    return newState;
  }

  return state;
};

export default editedTodos;
