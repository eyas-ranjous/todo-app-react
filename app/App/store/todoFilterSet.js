const todoFilterSet = (state = new Set(), action) => {
  if (action.type === 'UPDATE_STATUS_FILTER_SET') {
    return new Set(action.status);
  }

  return state;
};

export default todoFilterSet;
