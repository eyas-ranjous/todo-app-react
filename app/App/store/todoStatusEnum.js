export const TodoStatusEnum = {
  TODO: 1,
  DONE: 2
};

export const TodoStatus = (state = TodoStatusEnum) => state;
