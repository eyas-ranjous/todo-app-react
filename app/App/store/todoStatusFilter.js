const todoStatusFilter = (state = new Set(), action) => {
  if (action.type === 'SET_TODO_STATUS_FILTER') {
    return new Set(action.status);
  }

  return state;
};

export default todoStatusFilter;
