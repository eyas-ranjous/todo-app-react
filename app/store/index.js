import { createStore, combineReducers } from 'redux';
import addedTodos from './addedTodos';
import editedTodos from './editedTodos';
import { TodoStatus } from './todoStatusEnum';
import { TodoFilters } from './todoFiltersEnum';
import todoFilterSet from './todoFilterSet';
import selectedFilter from './selectedFilter';

const reducers = combineReducers({
  addedTodos,
  editedTodos,
  TodoStatus,
  TodoFilters,
  todoFilterSet,
  selectedFilter
});

export default createStore(reducers);
