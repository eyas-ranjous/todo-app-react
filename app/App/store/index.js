import { createStore, combineReducers } from 'redux';
import addedTodos from './addedTodos';
import editedTodos from './editedTodos';
import { TodoStatus } from './todoStatus';
import todoStatusFilter from './todoStatusFilter';

const reducers = combineReducers({
  addedTodos,
  editedTodos,
  TodoStatus,
  todoStatusFilter
});

export default createStore(reducers);
