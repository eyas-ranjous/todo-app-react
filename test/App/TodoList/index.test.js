import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux'
import { TodoStatusEnum } from '../../../app/store/TodoStatusEnum';
import createStore from '../../../app/store';
import TodoItem from '../../../app/App/TodoItem';
import TodoListWith from '../../../app/App/TodoList';

describe('TodoList', () => {
  const store = createStore();
  const TodoList = TodoListWith(TodoItem);

  it('renders successfully', () => {
    store.dispatch({ type: 'ADD_TODO', todo: { text: 'test task 1' } });
    store.dispatch({ type: 'ADD_TODO', todo: { text: 'test task 2' } });
    store.dispatch({ type: 'ADD_TODO', todo: { text: 'test task 3' } });
    const wrapper = mount(
      <Provider store={store}>
        <TodoList />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
