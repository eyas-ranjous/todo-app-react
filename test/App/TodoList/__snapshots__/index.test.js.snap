// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders successfully 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Component>
    <ul
      className="todo-list"
    >
      <TodoItem
        key="todo-1"
        todo={
          Object {
            "id": "todo-1",
            "status": 1,
            "text": "test task 1",
          }
        }
      >
        <li>
          <TodoStatusIcon
            todo={
              Object {
                "id": "todo-1",
                "status": 1,
                "text": "test task 1",
              }
            }
          >
            <span
              className="todo-item-option"
              onClick={[Function]}
            >
              ⬜
            </span>
          </TodoStatusIcon>
          <RemoveIcon
            todo={
              Object {
                "id": "todo-1",
                "status": 1,
                "text": "test task 1",
              }
            }
          >
            <span
              className="todo-item-option todo-item-remove"
              onClick={[Function]}
            >
              ✘
            </span>
          </RemoveIcon>
          <EditIcon
            todo={
              Object {
                "id": "todo-1",
                "status": 1,
                "text": "test task 1",
              }
            }
          >
            <span
              className="todo-item-option todo-item-edit"
              onClick={[Function]}
            >
              ✎
            </span>
          </EditIcon>
          <TodoText
            todo={
              Object {
                "id": "todo-1",
                "status": 1,
                "text": "test task 1",
              }
            }
          >
            <span
              className=""
            >
              test task 1
            </span>
          </TodoText>
        </li>
      </TodoItem>
      <TodoItem
        key="todo-2"
        todo={
          Object {
            "id": "todo-2",
            "status": 1,
            "text": "test task 2",
          }
        }
      >
        <li>
          <TodoStatusIcon
            todo={
              Object {
                "id": "todo-2",
                "status": 1,
                "text": "test task 2",
              }
            }
          >
            <span
              className="todo-item-option"
              onClick={[Function]}
            >
              ⬜
            </span>
          </TodoStatusIcon>
          <RemoveIcon
            todo={
              Object {
                "id": "todo-2",
                "status": 1,
                "text": "test task 2",
              }
            }
          >
            <span
              className="todo-item-option todo-item-remove"
              onClick={[Function]}
            >
              ✘
            </span>
          </RemoveIcon>
          <EditIcon
            todo={
              Object {
                "id": "todo-2",
                "status": 1,
                "text": "test task 2",
              }
            }
          >
            <span
              className="todo-item-option todo-item-edit"
              onClick={[Function]}
            >
              ✎
            </span>
          </EditIcon>
          <TodoText
            todo={
              Object {
                "id": "todo-2",
                "status": 1,
                "text": "test task 2",
              }
            }
          >
            <span
              className=""
            >
              test task 2
            </span>
          </TodoText>
        </li>
      </TodoItem>
      <TodoItem
        key="todo-3"
        todo={
          Object {
            "id": "todo-3",
            "status": 1,
            "text": "test task 3",
          }
        }
      >
        <li>
          <TodoStatusIcon
            todo={
              Object {
                "id": "todo-3",
                "status": 1,
                "text": "test task 3",
              }
            }
          >
            <span
              className="todo-item-option"
              onClick={[Function]}
            >
              ⬜
            </span>
          </TodoStatusIcon>
          <RemoveIcon
            todo={
              Object {
                "id": "todo-3",
                "status": 1,
                "text": "test task 3",
              }
            }
          >
            <span
              className="todo-item-option todo-item-remove"
              onClick={[Function]}
            >
              ✘
            </span>
          </RemoveIcon>
          <EditIcon
            todo={
              Object {
                "id": "todo-3",
                "status": 1,
                "text": "test task 3",
              }
            }
          >
            <span
              className="todo-item-option todo-item-edit"
              onClick={[Function]}
            >
              ✎
            </span>
          </EditIcon>
          <TodoText
            todo={
              Object {
                "id": "todo-3",
                "status": 1,
                "text": "test task 3",
              }
            }
          >
            <span
              className=""
            >
              test task 3
            </span>
          </TodoText>
        </li>
      </TodoItem>
    </ul>
  </Component>
</Provider>
`;
